iris-mpc-cpu/src/network/grpc.rs

```diff
@@ -0,0 +1,491 @@
+        Ok(())
+    }
+}
+
+// Server implementation
+#[async_trait]
+impl PartyNode for GrpcNetworking {
+    async fn send_message(
+        &self,
+        request: Request<SendRequest>,
+    ) -> TonicResult<Response<SendResponse>> {
+        let request = request.into_inner();
+        let sender_id = Identity::from(request.sender_id);
+        if sender_id == self.party_id {
+            return Err(Status::unauthenticated(format!(
+                "Sender ID coincides with receiver ID: {:?}",
+                sender_id
+            )));
+        }
+        let session_id = SessionId::from(request.session_id);
+        let message_queue = self
+            .message_queues
+            .get(&session_id)
+            .ok_or(Status::not_found(format!(
+                "Session {:?} hasn't been created",
+                session_id
+            )))?;
+        message_queue
+            .push_back(&sender_id, request.data)
+            .await
+            .map_err(err_to_status)?;
+        Ok(Response::new(SendResponse {}))
+    }
+}
+
+// Client implementation
+#[async_trait]
+impl Networking for GrpcNetworking {
+    async fn send(
+        &self,
+        value: Vec<u8>,
+        receiver: &Identity,
+        session_id: &SessionId,
+    ) -> eyre::Result<()> {
+        let backoff = ExponentialBackoff {
+            max_elapsed_time: Some(std::time::Duration::from_secs(2)),
+            max_interval: std::time::Duration::from_secs(1),
+            multiplier: 1.1,
+            ..Default::default()
+        };
+        retry(backoff, || async {
+            // Send message via gRPC client
+            let mut client = self
+                .clients
+                .get(receiver)
+                .ok_or_eyre(format!("Client not found {:?}", receiver))?
+                .clone();
+            let request = Request::new(SendRequest {
+                sender_id:  self.party_id.0.clone(),
+                session_id: session_id.0,
+                data:       value.clone(),
+            });
+            println!(
+                "Sending message {:?} from {:?} to {:?}",
+                value, self.party_id, receiver
+            );
+            let _response = client
+                .send_message(request)
+                .await
+                .map_err(|err| eyre!(err.to_string()))?;
+            println!(
+                "SUCCESS: Sending message {:?} from {:?} to {:?}",
+                value, self.party_id, receiver
+            );
+            Ok(())
+        })
+        .await
+    }
+
+    async fn receive(&self, sender: &Identity, session_id: &SessionId) -> eyre::Result<Vec<u8>> {
+        // Just retrieve the first message from the corresponding queue
+        self.message_queues
+            .get(session_id)
+            .ok_or_eyre(format!(
+                "Session {session_id:?} hasn't been added to message queues"
+            ))?
+            .pop_front(sender)
+            .await
+    }
```

Comment:
Does this get + pop_front combo hold the dashmap entry lock until there is actual data from the network?
I think this only works atm because the async channel used internally does not need any mut references to self at all, right, and would not work with e.g., a tokio channel

Reply:
You're correct. Do you see a case where it might fail in our protocols?

Reply:
I think it is fine due to the design of aysnc channel, but IDK the tradeoff that crate makes to be `&self` only compared to most channel impls that have `&mut self` signatures